### PR TITLE
[Feature] `files()` supports CSV format

### LINKED ISSUE #37922: `Files()` support CSV format
## Feature request
The `files()` only supports parquet and orc. The csv  format is not supported yet.

### PR DESCRIPTION
**Pull request by rickif**
Why I'm doing:
Now, we can use `files()` to load files from the storage system, but the csv format is not supported.
What I'm doing:
This PR adds the support of csv format to `files()`.
```
MySQL [db0]> select * from
 files("path"="file://xxx/csv_file23", "format"="csv", "csv.column_separator"=",", "auto_detect_sample_files"="1", "auto_detect_sample_rows"="1", "csv.enclose"='"', "csv.skip_header"="1", "csv.escape"="\\");
+------+-------+------+------+
| $1   | $2    | $3   | $4   |
+------+-------+------+------+
|    2 | Jok,e | 21.3 |    0 |
+------+-------+------+------+
1 row in set (0.289 sec)

```
This PR includes following changes:
1. add csv format and all csv options to `files()`
2. add `auto_detect_sample_rows` to control the sample rows of csv file
3. fill columns with null for not existing field

Fixes https://github.com/StarRocks/starrocks/issues/37922

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java [hunk 1]
```diff
@@ -190,7 +205,7 @@ private void parseProperties() throws DdlException {
             throw new DdlException("format is null. Please add properties(format='xxx') when create table");
         }
 
-        if (!format.equalsIgnoreCase("parquet") && !format.equalsIgnoreCase("orc")) {
+        if (!SUPPORTED_FORMATS.contains(format.toLowerCase())) {
```
Comment (jaogoy) [inline fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java:208]: Why not use getFileFormatType() first to convert `format` to TFileFormatType (as the standard internal format description), then check whether it's supported?

File: fe/fe-core/src/main/java/com/starrocks/load/Load.java [hunk 2]
```diff
@@ -1203,5 +1204,41 @@ public long saveLoadJob(DataOutputStream dos, long checksum) throws IOException
...
+            // Attention: The compression type of csv format is from the suffix of filename.
+        }
+
+        String lowerCasePath = path.toLowerCase();
+        if (lowerCasePath.endsWith(".parquet") || lowerCasePath.endsWith(".parq")) {
+            return TFileFormatType.FORMAT_PARQUET;
+        } else if (lowerCasePath.endsWith(".orc")) {
+            return TFileFormatType.FORMAT_ORC;
+        } else if (lowerCasePath.endsWith(".gz")) {
+            return TFileFormatType.FORMAT_CSV_GZ;
```
Comment (jaogoy) [inline fe/fe-core/src/main/java/com/starrocks/load/Load.java:1230]: We will support COMPRESSION for CSV, JSON, Avro and other file formats in the near steps, How can we support it?

Comment (jaogoy) [reply fe/fe-core/src/main/java/com/starrocks/load/Load.java:1230]: In my opinion, we can directly use FileFormat and compression type separately in FILES, regardless of it in Stream Load.

File: be/src/exec/file_scanner.cpp [hunk 3]
```diff
@@ -443,6 +443,10 @@ Status FileScanner::sample_schema(RuntimeState* state, const TBrokerScanRange& s
             p_scanner = std::make_unique<ORCScanner>(state, &profile, sample_range, &counter, true);
             break;
 
+        case TFileFormatType::FORMAT_CSV_PLAIN:
```
Comment (jaogoy) [inline be/src/exec/file_scanner.cpp:446]: So, currently , SR doesn't support compression when file format is CSV, right?

File: be/src/exec/csv_scanner.cpp [hunk 4]
```diff
@@ -506,4 +514,45 @@ void CSVScanner::_report_rejected_record(const CSVReader::Record& record, const
     _state->append_rejected_record_to_file(record.to_string(), err_msg, _curr_reader->filename());
 }
 
+static TypeDescriptor get_type_desc(const Slice& field) {
+    StringParser::ParseResult result;
+
+    StringParser::string_to_int<int64_t>(field.get_data(), field.get_size(), &result);
```
Comment (jaogoy) [inline be/src/exec/csv_scanner.cpp:520]: Will it return SUCCESS to parse `123.4` to an int64_t?

File: be/src/exec/csv_scanner.cpp [hunk 5]
```diff
@@ -506,4 +514,45 @@ void CSVScanner::_report_rejected_record(const CSVReader::Record& record, const
...
+    if (result == StringParser::PARSE_SUCCESS) {
+        return TypeDescriptor(TYPE_BIGINT);
+    }
+
+    StringParser::string_to_float<double>(field.get_data(), field.get_size(), &result);
```
Comment (jaogoy) [inline be/src/exec/csv_scanner.cpp:525]: It's OK to use simple type detection. we can refine it in the future.

File: be/src/exec/csv_scanner.cpp [hunk 6]
```diff
@@ -506,4 +514,45 @@ void CSVScanner::_report_rejected_record(const CSVReader::Record& record, const
...
+    StringParser::string_to_bool(field.get_data(), field.get_size(), &result);
+    if (result == StringParser::PARSE_SUCCESS) {
+        return TypeDescriptor(TYPE_BOOLEAN);
+    }
+
+    // default VARCHAR.
+    return TypeDescriptor::create_varchar_type(TypeDescriptor::MAX_VARCHAR_LENGTH);
+}
+
+Status CSVScanner::get_schema(std::vector<SlotDescriptor>* schema) {
```
Comment (jaogoy) [inline be/src/exec/csv_scanner.cpp:539]: I'm curiouse why we like using pointer instead of reference?

File: be/src/exec/csv_scanner.cpp [hunk 7]
```diff
@@ -506,4 +514,45 @@ void CSVScanner::_report_rejected_record(const CSVReader::Record& record, const
...
+    if (schema == nullptr) {
+        return Status::InternalError("ouput schema is null");
+    }
+
+    RETURN_IF_ERROR(_init_reader());
+
+    CSVReader::Record record;
+    RETURN_IF_ERROR(_curr_reader->next_record(&record));
```
Comment (jaogoy) [inline be/src/exec/csv_scanner.cpp:547]: Read only one record? Why not merge several record-schema?

File: be/src/exec/csv_scanner.cpp [hunk 8]
```diff
@@ -506,4 +514,45 @@ void CSVScanner::_report_rejected_record(const CSVReader::Record& record, const
...
+
+    CSVReader::Fields fields;
+    _curr_reader->split_record(record, &fields);
+    for (size_t i = 0; i < fields.size(); i++) {
+        // column name: $1, $2, $3...
+        schema->emplace_back(SlotDescriptor(i, fmt::format("${}", i + 1), get_type_desc(fields[i])));
```
Comment (jaogoy) [inline be/src/exec/csv_scanner.cpp:553]: Put the magic const string into the csv header file, like  DEFAULT_CSV_COLUMN_NAME_FORMAT = "${}" ?

Comment (jaogoy) [reply be/src/exec/csv_scanner.cpp:553]: And, where is the code to merge several sample file schema?

Comment (rickif) [reply be/src/exec/csv_scanner.cpp:553]: https://github.com/StarRocks/starrocks/pull/36892

File: be/src/exec/csv_scanner.cpp (hunk 7 above)
Comment (rickif) [reply be/src/exec/csv_scanner.cpp:547]: How to control the number of lines to sample?

File: be/src/exec/csv_scanner.cpp (hunk 6 above)
Comment (rickif) [reply be/src/exec/csv_scanner.cpp:539]: It is the recommended code style.

File: be/src/exec/csv_scanner.cpp (hunk 4 above)
Comment (rickif) [reply be/src/exec/csv_scanner.cpp:520]: No.

File: fe/fe-core/src/main/java/com/starrocks/load/Load.java (hunk 2 above)
Comment (rickif) [reply fe/fe-core/src/main/java/com/starrocks/load/Load.java:1230]: Now, we get the compression type from the file name suffix. Maybe we could use a specialized property like `compression_type` to represent this.

File: fe/fe-core/src/main/java/com/starrocks/load/BrokerFileGroup.java [hunk 9]
```diff
@@ -137,8 +137,8 @@ public BrokerFileGroup(TableFunctionTable table) throws AnalysisException {
         this.filePaths.add(table.getPath());
 
         this.fileFormat = table.getFormat();
-        this.columnSeparator = "\t";
-        this.rowDelimiter = "\n";
+        this.columnSeparator = table.getCsvOptions().columnSeparator;
```
[the remark this answers is not in the document]
Comment (rickif) [reply fe/fe-core/src/main/java/com/starrocks/load/BrokerFileGroup.java:140]: `rowDelimiter` and `columnSeparator` already exist.

File: fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java [hunk 10]
```diff
@@ -76,6 +85,10 @@ public class TableFunctionTable extends Table {
 
     public static final String PROPERTY_AUTO_DETECT_SAMPLE_FILES = "auto_detect_sample_files";
 
+    public static final String PROPERTY_CSV_COLUMN_SEPARATOR = "csv.column_separator";
```
[the remark this answers is not in the document]
Comment (rickif) [reply fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java:88]: 1. Currently, it only appears in `files()`

File: fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java (hunk 1 above)
Comment (rickif) [reply fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java:208]: The `getFileFormatType` is used by broker load and `files`, but this check is only available in `files`.

File: be/src/exec/csv_scanner.cpp [hunk 11]
```diff
@@ -506,4 +514,45 @@ void CSVScanner::_report_rejected_record(const CSVReader::Record& record, const
...
+
+    RETURN_IF_ERROR(_init_reader());
+
+    CSVReader::Record record;
+    RETURN_IF_ERROR(_curr_reader->next_record(&record));
+
+    CSVReader::Fields fields;
+    _curr_reader->split_record(record, &fields);
+    for (size_t i = 0; i < fields.size(); i++) {
+        // column name: $1, $2, $3...
```
Comment (wyb) [inline be/src/exec/csv_scanner.cpp:552]: why not start from $0?

File: be/src/exec/csv_scanner.cpp (hunk 7 above)
Comment (wyb) [reply be/src/exec/csv_scanner.cpp:547]: maybe end of file or empty record

File: be/src/exec/csv_scanner.cpp [hunk 12]
```diff
@@ -506,4 +514,97 @@ void CSVScanner::_report_rejected_record(const CSVReader::Record& record, const
...
+        return Status::InternalError("ouput schema is null");
+    }
+
+    RETURN_IF_ERROR(_init_reader());
+
+    CSVReader::Record record;
+    // skip empty record.
+    do {
+        auto st = _curr_reader->next_record(&record);
+        if (!st.ok() && st.is_end_of_file()) {
```
Comment (wyb) [inline be/src/exec/csv_scanner.cpp:556]:
```suggestion
        if (st.is_end_of_file()) {
```

File: be/src/exec/csv_scanner.cpp [hunk 13]
```diff
@@ -506,4 +514,97 @@ void CSVScanner::_report_rejected_record(const CSVReader::Record& record, const
...
+        } else if (!st.ok()) {
+            return st;
+        }
+    }
+
+    CSVRow row;
+    // skip empty row.
+    do {
+        auto st = _curr_reader->next_record(row);
+        if (!st.ok() && st.is_end_of_file()) {
```
Comment (wyb) [inline be/src/exec/csv_scanner.cpp:594]:
```suggestion
        if (st.is_end_of_file()) {
```

File: be/src/exec/csv_scanner.cpp [hunk 14]
```diff
@@ -506,4 +514,97 @@ void CSVScanner::_report_rejected_record(const CSVReader::Record& record, const
...
+            return Status::OK();
+        } else if (!st.ok()) {
+            return st;
+        }
+        while (row.columns.empty())
+            ;
```
Comment (wyb) [inline be/src/exec/csv_scanner.cpp:600]: bug

File: fe/fe-core/src/main/java/com/starrocks/planner/FileScanNode.java [hunk 15]
```diff
@@ -160,6 +160,7 @@ public int compare(TBrokerFileStatus o1, TBrokerFileStatus o2) {
     private boolean useVectorizedLoad;
 
     private LoadJob.JSONOptions jsonOptions = new LoadJob.JSONOptions();
+    private boolean flexibleColumnMapping;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/planner/FileScanNode.java:163]:
```suggestion
    private boolean flexibleColumnMapping = false;
```

File: be/src/exec/csv_scanner.cpp [hunk 16]
```diff
@@ -452,11 +486,36 @@ Status CSVScanner::_parse_csv(Chunk* chunk) {
...
+                                _num_fields_in_csv, fields.size(), _parse_options);
+                        _report_error(record, error_msg);
+                    }
+                    if (_state->enable_log_rejected_record()) {
+                        std::string error_msg = make_column_count_not_matched_error_message(
+                                _num_fields_in_csv, fields.size(), _parse_options);
+                        _report_rejected_record(record, error_msg);
+                    }
+                    error_reported = true;
+                }
```
Comment (wyb) [inline be/src/exec/csv_scanner.cpp:515]: k++;

File: be/src/exec/csv_scanner.cpp [hunk 17]
```diff
@@ -360,11 +368,37 @@ Status CSVScanner::_parse_csv_v2(Chunk* chunk) {
...
+                        _report_error(record, error_msg);
+                    }
+                    if (_state->enable_log_rejected_record()) {
+                        std::string error_msg = make_column_count_not_matched_error_message(
+                                _num_fields_in_csv, row.columns.size(), _parse_options);
+                        _report_rejected_record(record, error_msg);
+                    }
+                }
+
+                error_reported = true;
```
Comment (wyb) [inline be/src/exec/csv_scanner.cpp:398]: move `error_reported = true;` in if {}

Comment (wyb) [reply be/src/exec/csv_scanner.cpp:398]: k++;

**Review by wyb — DISMISSED** [review]

**Review by wyb — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Review by nshangyiming — APPROVED** [review]
LGTM

**Pull request merged**